File: src/io/serial_manager.py
import serial
import serial.tools.list_ports
from src.io.io_transport import IOTransport
from src.io.serial_reader import SerialReaderThread

_STOPBITS_MAP = {
    1: serial.STOPBITS_ONE,
    1.5: serial.STOPBITS_ONE_POINT_FIVE,
    2: serial.STOPBITS_TWO,
}
_PARITY_MAP = {
    'None': serial.PARITY_NONE,
    'Even': serial.PARITY_EVEN,
    'Odd': serial.PARITY_ODD,
    'Mark': serial.PARITY_MARK,
    'Space': serial.PARITY_SPACE,
}


class SerialManager(IOTransport):

    def __init__(self):
        super().__init__()
        self.serial = serial.Serial()
        self.settings = {
            'baudrate': 115200,
            'databits': 8,
            'stopbits': 1,
            'parity': 'None',
            'flowcontrol': 'None',
            'frame_timeout': 50,
        }

    def get_available_ports(self):
        ports = serial.tools.list_ports.comports()
        return [(p.device, p.description) for p in ports]

    def get_available_devices(self):
        return self.get_available_ports()

    def _apply_serial_params(self):
        """将 settings 中的参数应用到 serial 对象。"""
        self.serial.baudrate = self.settings['baudrate']
        self.serial.bytesize = self.settings['databits']
        self.serial.stopbits = _STOPBITS_MAP.get(self.settings['stopbits'], serial.STOPBITS_ONE)
        self.serial.parity = _PARITY_MAP.get(self.settings['parity'], serial.PARITY_NONE)

    def reconfigure(self):
        with self._lock:
            if not self.is_connected or not self.serial.is_open:
                return False
            try:
                self._apply_serial_params()
                return True
            except Exception as e:
                self.error_occurred.emit(f"重新配置失败: {str(e)}")
                return False

    def _connect_impl(self, port):
        self.serial.port = port
        self._apply_serial_params()
        self.serial.open()
        try:
            frame_timeout = self.settings.get('frame_timeout', 50)
            thread = SerialReaderThread(self.serial, frame_timeout)
            self._connect_reader_thread(thread)
        except Exception:
            self.serial.close()
            raise

    def _close_resource(self):
        if self.serial.is_open:
            self.serial.close()

    def _send_bytes(self, data: bytes) -> bool:
        self.serial.write(data)
        return True

File: src/io/serial_manager.cpp
#include "serial_manager.h"
#include "serial_reader.h"

SerialManager::SerialManager(QObject *parent)
    : IOTransport(parent), m_serial(new QSerialPort(this))
{
}

QList<QPair<QString, QString>> SerialManager::getAvailablePorts() {
    QList<QPair<QString, QString>> ports;
    for (const QSerialPortInfo &info : QSerialPortInfo::availablePorts()) {
        ports.append({info.portName(), info.description()});
    }
    return ports;
}

QList<QPair<QString, QString>> SerialManager::getAvailableDevices() {
    return getAvailablePorts();
}

bool SerialManager::connectImpl(const QVariantMap &params) {
    QString port = params.value("port", m_settings.value("port")).toString();
    if (port.isEmpty()) {
        emit errorOccurred("No port specified");
        return false;
    }

    m_serial->setPortName(port);
    applySerialParams();

    if (!m_serial->open(QIODevice::ReadWrite)) {
        emit errorOccurred("Cannot open port: " + m_serial->errorString());
        return false;
    }

    // Create reader thread
    int frameTimeout = m_settings.value("frame_timeout", 50).toInt();
    auto *reader = new SerialReaderThread(m_serial, frameTimeout);
    connect(reader, &SerialReaderThread::dataReceived, this, &IOTransport::dataReceived);
    connect(reader, &SerialReaderThread::errorOccurred, this, &IOTransport::errorOccurred);
    startReaderThread(reader);

    return true;
}

void SerialManager::closeResource() {
    if (m_serial->isOpen()) {
        m_serial->close();
    }
}

bool SerialManager::sendBytes(const QByteArray &data) {
    if (!m_serial->isOpen()) return false;
    qint64 written = m_serial->write(data);
    return written == data.size();
}

void SerialManager::applySerialParams() {
    // Baudrate
    int baudrate = m_settings.value("baudrate", 115200).toInt();
    m_serial->setBaudRate(baudrate);

    // Data bits
    int databits = m_settings.value("databits", 8).toInt();
    switch (databits) {
    case 5: m_serial->setDataBits(QSerialPort::Data5); break;
    case 6: m_serial->setDataBits(QSerialPort::Data6); break;
    case 7: m_serial->setDataBits(QSerialPort::Data7); break;
    default: m_serial->setDataBits(QSerialPort::Data8); break;
    }

    // Stop bits
    float stopbits = m_settings.value("stopbits", 1.0).toFloat();
    if (stopbits == 1.5f) m_serial->setStopBits(QSerialPort::OneAndHalfStop);
    else if (stopbits == 2.0f) m_serial->setStopBits(QSerialPort::TwoStop);
    else m_serial->setStopBits(QSerialPort::OneStop);

    // Parity
    QString parity = m_settings.value("parity", "None").toString();
    if (parity == "Even") m_serial->setParity(QSerialPort::EvenParity);
    else if (parity == "Odd") m_serial->setParity(QSerialPort::OddParity);
    else if (parity == "Mark") m_serial->setParity(QSerialPort::MarkParity);
    else if (parity == "Space") m_serial->setParity(QSerialPort::SpaceParity);
    else m_serial->setParity(QSerialPort::NoParity);

    // Flow control
    QString fc = m_settings.value("flowcontrol", "None").toString();
    if (fc == "RTS-CTS") m_serial->setFlowControl(QSerialPort::HardwareControl);
    else if (fc == "XON-XOFF") m_serial->setFlowControl(QSerialPort::SoftwareControl);
    else m_serial->setFlowControl(QSerialPort::NoFlowControl);
}

void SerialManager::reconfigure() {
    if (m_serial->isOpen()) {
        applySerialParams();
    }
}
